cpp/include/cudf/detail/scan_reduce_iterator.cuh

```diff
@@ -0,0 +1,267 @@
+  {
+    --itr_;
+    return *this;
+  }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator operator--(int)
+  {
+    scan_reduce_iterator tmp(*this);
+    operator--();
+    return tmp;
+  }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator& operator+=(difference_type offset)
+  {
+    itr_ += offset;
+    return *this;
+  }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator operator+(difference_type offset) const
+  {
+    scan_reduce_iterator tmp(*this);
+    tmp.itr_ += offset;
+    return tmp;
+  }
+
+  CUDF_HOST_DEVICE inline friend scan_reduce_iterator operator+(difference_type offset,
+                                                                scan_reduce_iterator const& rhs)
+  {
+    scan_reduce_iterator tmp{rhs};
+    tmp.itr_ += offset;
+    return tmp;
+  }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator& operator-=(difference_type offset)
+  {
+    itr_ -= offset;
+    return *this;
+  }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator operator-(difference_type offset) const
+  {
+    scan_reduce_iterator tmp(*this);
+    tmp.itr_ -= offset;
+    return tmp;
+  }
+
+  CUDF_HOST_DEVICE inline friend scan_reduce_iterator operator-(difference_type offset,
+                                                                scan_reduce_iterator const& rhs)
+  {
+    scan_reduce_iterator tmp{rhs};
+    tmp.itr_ -= offset;
+    return tmp;
+  }
+
+  CUDF_HOST_DEVICE inline difference_type operator-(scan_reduce_iterator const& rhs) const
+  {
+    return itr_ - rhs.itr_;
+  }
+  CUDF_HOST_DEVICE inline bool operator==(scan_reduce_iterator const& rhs) const
+  {
+    return rhs.itr_ == itr_;
+  }
+  CUDF_HOST_DEVICE inline bool operator!=(scan_reduce_iterator const& rhs) const
+  {
+    return rhs.itr_ != itr_;
+  }
+  CUDF_HOST_DEVICE inline bool operator<(scan_reduce_iterator const& rhs) const
+  {
+    return itr_ < rhs.itr_;
+  }
+  CUDF_HOST_DEVICE inline bool operator>(scan_reduce_iterator const& rhs) const
+  {
+    return itr_ > rhs.itr_;
+  }
+  CUDF_HOST_DEVICE inline bool operator<=(scan_reduce_iterator const& rhs) const
+  {
+    return itr_ <= rhs.itr_;
+  }
+  CUDF_HOST_DEVICE inline bool operator>=(scan_reduce_iterator const& rhs) const
+  {
+    return itr_ >= rhs.itr_;
+  }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator const& operator*() const { return *this; }
+
+  CUDF_HOST_DEVICE inline scan_reduce_iterator const operator[](int idx) const
+  {
+    scan_reduce_iterator tmp{*this};
+    tmp.itr_ += idx;
+    return tmp;
+  }
+
+  /**
+   * @brief Called to set the output of the scan operation to the current iterator position
+   *
+   * @param value Value to set to the current output
+   * @return This iterator instance
+   */
+  CUDF_HOST_DEVICE inline scan_reduce_iterator const& operator=(ReduceType const value) const
+  {
+    *itr_ = static_cast<ScanType>(value);  // place into the output
+    if (itr_ == end_) { *last_ = value; }  // also save the last value
```

Comment:
I may be misunderstanding, but shouldn't this line be `*last_ += value` to perform the sum-reduction?